apps/safety/main.cpp

```diff
@@ -1,13 +1,219 @@
+        {
+            led1 = 1;
+        }
+
+        float measuredVoltage = getVoltageData();
+        if (inRange(voltageUpperLim, voltageLowerLim, measuredVoltage))
+        {
+            led2 = 1;
+        }
+
+        float measuredTemp = readThermosistor();
+        if(inRange(tempUpperLim, tempLowerLim, measuredTemp))
+        {
+            led3 = 1;
+        }
     }
 }
+
+void initCAN()
+{
+    // will need new id's 
+    can.filter(CAN1_RX, CAN1_TX, ROVER_CANID_MASK, CANStandard);
+}
+
+void processCANMsg(CANMsg *p_newMsg)
+{
+    // PRINT_INFO("Recieved CAN message with ID %X\r\n", p_newMsg->id);
+    // The specific can ID for configuring chip is ___
+    switch (p_newMsg->id){
+        case expectedCANID:
+            pc.printf("Updating INA226 config\r\n");
+            break;
+        default:
+            pc.printf("Recieved unimplemented command\r\n");
+            break;
+    }
+}
+
+
+int inRange(float upperLim, float lowerLim, float val)
+{
+    int inRange = (val > upperLim || val< lowerLim) ? 1: 0;
+    return inRange;
+}
+
+
+float getCurrentData()
+{
+    //some conversion factor for current data
+    float conversionFactor = 1;
+    //may need char16_t?
+    char cmd[3] = {CURRENT_REGISTER};
+    
+    i2c.write(SENSOR_SLAVE_ADDRESS, cmd, 2);
+    i2c.read(SENSOR_SLAVE_ADDRESS, cmd, 2, false);
+    
+    char16_t currentData = cmd[1]; 
+    currentData = (cmd[2] << 8 | 0x00);
+    
+    return currentData * conversionFactor;
+}
+
+
+float getVoltageData()
+{
+    //some conversion factor for voltage data
+    float conversionFactor = 1;
+    char cmd[3] = {VOLTAGE_REGISTER};
+    
+    i2c.write(SENSOR_SLAVE_ADDRESS, cmd, 2);
+    i2c.read(SENSOR_SLAVE_ADDRESS, cmd, 2, false);
+    
+    char16_t voltageData = cmd[1]; 
+    voltageData = (cmd[2] << 8 | 0x00);
+    
+    return voltageData * conversionFactor;
+}
+
+
+float readThermosistor(float vin, float vout)
+{
+    //voltage divider
+    float rthermo = rResistor *((vin/vout)-1);
+
+    //convert resistance to temp (steinhart and hart eqn)
+    return 1/(A + B * log(rthermo) + C * pow(log(rthermo),3));
+}
+
+int configureSensor(configurationMode operationMode, int resetRegisters)
+{
+    char cmd[3] = {CONFIG_REGISTER, 0x00, 0x00};
+    int dataByte = 0x00;
+
+    //fill up register with configuration data
+    switch (operationMode)
+    {
+    case powerReset:
+        dataByte |= powerReset;
+        break;
+    case continous:
+        dataByte |= continous;
+        break;
+    case triggered:
+        dataByte |= triggered;
+        break;
+    default:
+        break;
+    }
+    
+    // averaging mode default is 0 so not setting is needed
+    dataByte |= (shuntVoltageConversionTime << 3);
+    dataByte |= (busVoltageConversionTime << 6);
+    
+    if (resetRegisters)
+    {
+        dataByte |= (resetRegisters << 15);
+    }
+    
+    // give cmd the data to write 
+    cmd[1] |= dataByte;
```

Comment:
there is no point in using |=, should just use =; also assigning an int that holds a value larger than what can be held in a char to a char results in undefined behavior. A better way would be cmd[1] = dataByte & 0xFF;